### DISCUSSION TITLE
Hausdorff Distance metric gets 'inf'

### DISCUSSION THREAD

**Discussion by hsiangyuzhao** [Project-MONAI/MONAI#2179] [Q&A]
Hi, thank you guys for developing such a good package for medical image computation. I am using MONAI to perform lesion segmentation on BraTS dataset and I use Dice and **Hausdorff distance** for metrics. But when training I found that **for some certain cases the hausdorff distance would be inf (not nan)**, and I have no idea the cause of it. Would you guys help me to find out the cause?
**Relevant code is pasted below:**
![Snipaste_2021-05-12_15-20-22](https://user-images.githubusercontent.com/53631393/117935153-047f7580-b336-11eb-8e6b-c3107e5f2fbd.png)
as BraTS provides labels for enhancing tumor, NET/NCR and peritumoral edema, I need to transform these discrete labels into overrlaping ones (enhancing tumor, tumor core and whole tumor), the relevant code is pasted below:
![Snipaste_2021-05-12_15-20-43](https://user-images.githubusercontent.com/53631393/117935464-50321f00-b336-11eb-849c-393fbaa3623e.png)

**Comment by Nic-Ma** [comment]
Hi @hsiangyuzhao ,

Thanks for your interest here.
MONAI provides a tutorial that train segmentation task with Brats dataset:
https://github.com/Project-MONAI/tutorials/blob/master/3d_segmentation/brats_segmentation_3d.ipynb
Maybe you can compare it with your program first?
If still has the issue, please paste some training log then @yiheng-wang-nv can help try to figure it out.

Thanks.

**Reply by hsiangyuzhao** [reply]
Thanks for the reply. I have read the tutorials, but the **tutorials do not provide an example of the usage of Hausdorff distance metric**. In fact I did a coarse experiment of Hausdorff distance metric, the log is pasted below:
```python
>>> y_pred = torch.as_tensor([0,1,0,1,1,0],dtype=torch.float).reshape((1, 1, 2, 3))
>>> y_pred
tensor([[[[0., 1., 0.],
          [1., 1., 0.]]]])
>>> y = torch.as_tensor([0,0,0,0,0,0],dtype=torch.float).reshape((1, 1, 2, 3))
>>> y
tensor([[[[0., 0., 0.],
          [0., 0., 0.]]]])
>>> hausdorff_metric = HausdorffDistanceMetric(include_background=True, reduction='none', percentile=95)
>>> hausdorff_metric(y_pred, y)
(tensor([[inf]], dtype=torch.float64), tensor([[1.]]))
```
It seems that for a certain subject, **if the subject do not have certain label but the model predicts that it has, the hausdorff distance would be inf**. I am not that familiar with hausdorff distance metric, and **I wonder such behavior is the desired behavior**?

**Reply by Nic-Ma** [reply]
Hi @yiheng-wang-nv and @rijobro ,

Could you please help double confirm the behavior of `Hausdorff distance`?

Thanks.

**Reply by rijobro** [reply]
Hi @hsiangyuzhao , the Hausdorff distance compares two volumes. In your example, `y` is all zeroes, so I'm not overly surprised that the metric came out as `inf`. Similarly, if your network was also predicting all zeroes for certain segmentations, perhaps that would result in the same problem. 

It might be a good idea to find a pair of (`y, y_pred`) BRATS images that are giving you infinite Haussdorf distance and then we can try to figure out why. Maybe `y_pred` is all zeroes, maybe there's no overlap between the two volumes (I can't remember off the top of my head what the Hausdorff distance for that would be), maybe it's a bug in our code... :)

**Reply by hsiangyuzhao** [reply]
Hi @rijobro , thanks for your advice, I think it is wise to find out the volume that always produces the inf Hausdorff distance. I am working on it. Meanwhile, I made yet another coarse experiment of Hausdorff distance metric, the log is pasted below:
```python
>>> y_pred = torch.as_tensor([0,1,0,1,1,0],dtype=torch.float).reshape((1, 1, 2, 3))
>>> y_pred
tensor([[[[0., 1., 0.],
          [1., 1., 0.]]]])
>>> y = torch.as_tensor([0,0,0,0,0,0],dtype=torch.float).reshape((1, 1, 2, 3))
>>> y
tensor([[[[0., 0., 0.],
          [0., 0., 0.]]]])
>>> hausdorff_metric = HausdorffDistanceMetric(include_background=True, reduction='none', percentile=95)
>>> hausdorff_metric(y_pred, y)
(tensor([[inf]], dtype=torch.float64), tensor([[1.]]))
>>> # change y_pred from random values to all zeros, while y remains all zeros
>>> y_pred = torch.as_tensor([0,0,0,0,0,0],dtype=torch.float).reshape((1, 1, 2, 3))
>>> hausdorff_metric(y_pred, y)
(tensor([[nan]], dtype=torch.float64), tensor([[0.]]))
```
It seems that, on condition that target is all zeros, **when the network is predicting all zeros, it gets a *nan* result rather than *inf*; when the network is predicting some ones in the prediction, it gets an *inf*.** So I think it is probably **some volume has no enhancing tumor (the target is all zeros in the corresonding channel) but the network predicts it has, which leads to the inf result**. To my knowledge it is the most plausible case. Anyway I am still looking for the volume with problem and share it with you guys. I will reply soon as long as I find the black sheep :)

**Reply by hsiangyuzhao** [reply]
Hi @rijobro , I have found the potential cause. Just as I suspected, for some certain subjects (e.g. BraTS20_Training_329 in BraTS 2020 dataset), **the subject has no enhancing tumor labels, which means the corresponding channel in the y_true is all zeros**. And **if the model prediction in the corresponding channel is not all zeros, the Hausdorff distance metric would return an 'inf'**.
I am not familiar with Hausdorff distance metric, and I wonder whether this is the expected behavior? To my knowledge I think the metric returning a 'nan' rather than 'inf' would be more reasonable.

**Reply by yiheng-wang-nv** [reply]
Hi @hsiangyuzhao , according to the definition of Hausdorff distance, in your case, it should be "inf".
Here are some references:
https://www2.math.upenn.edu/~brweber/Lectures/USTC2011/Lecture%205%20-%20Gromov-Hausdorff%20distance.pdf

Deepmind's implementation:
https://github.com/deepmind/surface-distance/blob/master/surface_distance/metrics.py#L351

It seems I forgot to add the reference in the doc string, let me submit a PR to add it. Thanks!

**Reply by rijobro** [reply]
@hsiangyuzhao I think the `nan` is also creeping in because you use 95 percentile (bottom half of these results will use 95 percentile as opposed to default `None`):

```python
In [20]: for percentile in (None, 95):
    ...:     for include_background in (True, False):
    ...:         for A in (torch.zeros, torch.ones):
    ...:             for B in (torch.zeros, torch.ones):
    ...:                 a = A((1,1,20,20,20))
    ...:                 b = B((1,1,20,20,20))
    ...:                 print(HausdorffDistanceMetric(include_background, reduction='none', percentile=percentile)(a, b))
    ...: 
(tensor([[nan]], dtype=torch.float64), tensor([[0.]]))
(tensor([[inf]], dtype=torch.float64), tensor([[1.]]))
(tensor([[inf]], dtype=torch.float64), tensor([[1.]]))
(tensor([[0.]], dtype=torch.float64), tensor([[1.]]))
(tensor([[nan]], dtype=torch.float64), tensor([[0.]]))
(tensor([[inf]], dtype=torch.float64), tensor([[1.]]))
(tensor([[inf]], dtype=torch.float64), tensor([[1.]]))
(tensor([[0.]], dtype=torch.float64), tensor([[1.]]))
(tensor([[nan]], dtype=torch.float64), tensor([[0.]]))
invalid value encountered in subtract
(tensor([[nan]], dtype=torch.float64), tensor([[0.]]))
(tensor([[nan]], dtype=torch.float64), tensor([[0.]]))
(tensor([[0.]], dtype=torch.float64), tensor([[1.]]))
(tensor([[nan]], dtype=torch.float64), tensor([[0.]]))
(tensor([[nan]], dtype=torch.float64), tensor([[0.]]))
(tensor([[nan]], dtype=torch.float64), tensor([[0.]]))
(tensor([[0.]], dtype=torch.float64), tensor([[1.]]))
```

**Reply by rijobro** [reply]
@yiheng-wang-nv do you think the above results have too much variability? If so, we might want to put in some extra checks to the computation.

**Reply by hsiangyuzhao** [reply]
Hi @rijobro , I run the code you pasted above but get a different result, and I have no idea about the reason.
![Snipaste_2021-05-12_20-01-35](https://user-images.githubusercontent.com/53631393/117971846-f7c24800-b35c-11eb-8fed-8bd7295e81cc.png)
Could you please look into the reason of it?

**Reply by yiheng-wang-nv** [reply]
Hi @rijobro , could you please share the running environment of the above results? I got the same output as @hsiangyuzhao , which should be expected: the Hausdorff Distance between all ones tensor and all zeros tensor is inf.

**Reply by rijobro** [reply]
I have 3 machines that I've tested this on. Two gave me my initial result, one gave me results that matched yours. I think it's possibly down to scipy versions. The two that gave my initial result are on 1.6.3, whereas the one that gave results that matched those of @hsiangyuzhao is on version 1.5.2. What about you guys?

**Reply by hsiangyuzhao** [reply]
Hi @rijobro , my scipy version is 1.5.0.

**Reply by rijobro** [reply]
Ok, so I suspect it's due to the different scipy versions. Anyway, we can make things easier for the user by adding in these sorts of checks:

```python
if torch.all(y_pred == 0):
    warnings.warn("y_pred is all 0, this may result in nan/inf distance")
if torch.all(y == 0):
    warnings.warn("y is all 0, this may result in nan/inf distance")
```

**Reply by yiheng-wang-nv** [reply]
Hi @rijobro , thanks for the advice, I submitted a PR (https://github.com/Project-MONAI/MONAI/pull/2205) to add the warnings.

**Reply by umgpy** [reply]
Hi @rijobro 
I am trying to use Dice loss along with dice metric, but at the same time I would also like to use HD95 and average surface distance as an additional metric for reporting.
However, after a couple of epochs both HD95 and ASSD start to give Inf and the loss ends up being Nan. 
While on training DSC as a metric there is no such behavior.

code
```python
hd95_metric = HausdorffDistanceMetric(include_background=True, percentile=95, reduction="mean")
hd95_metric_batch = HausdorffDistanceMetric(include_background=True, percentile=95, reduction="mean_batch")

sd_metric =  SurfaceDistanceMetric(include_background=True, symmetric=True, reduction="mean")
sd_metric_batch =  SurfaceDistanceMetric(include_background=True, symmetric=True, reduction="mean_batch")`

...

model.eval()
        with torch.no_grad():

            for val_data in val_loader:
                ...
                val_outputs = inference(val_inputs)
                val_outputs = [post_trans(i) for i in decollate_batch(val_outputs)]
                ## trying to compute all three metrics but or this run only had SurfaceDIstance 
                #dice_metric(y_pred=val_outputs, y=val_labels)
                #dice_metric_batch(y_pred=val_outputs, y=val_labels)
                #hd95_metric(y_pred=val_outputs, y=val_labels)
                #hd95_metric_batch(y_pred=val_outputs, y=val_labels)
                sd_metric(y_pred=val_outputs, y=val_labels)
                sd_metric_batch(y_pred=val_outputs, y=val_labels)      
 
                assd_metric = sd_metric.aggregate().item()
                assd_metric_values.append(assd_metric)
                assd_metric_batch = sd_metric_batch.aggregate()
                assd_metric_1 = assd_metric_batch[0].item()
                assd_metric_values_1.append(assd_metric_1)
                assd_metric_2 = assd_metric_batch[1].item()
                assd_metric_values_2.append(assd_metric_2)
                assd_metric_3 = assd_metric_batch[2].item()
                assd_metric_values_3.append(assd_metric_3)
                sd_metric.reset()
                sd_metric_batch.reset()

```
epoch 10/300
1/8, train_loss: 0.5534, step time: 1.5471
2/8, train_loss: 0.5787, step time: 1.5402
3/8, train_loss: 0.5726, step time: 1.5288
4/8, train_loss: 0.5950, step time: 1.5261
5/8, train_loss: 0.5621, step time: 1.5294
6/8, train_loss: 0.5870, step time: 1.5272
7/8, train_loss: 0.5832, step time: 1.5286
8/8, train_loss: 0.5712, step time: 1.5279
epoch 10 average loss: 0.5754
current epoch: 10 current mean assd: 21.4569 BG: 10.9622 fem: 29.2373 pel: 24.1713
best mean dice: -1.0000 at epoch: -1
time consuming of epoch 10 is: 1241.7538

epoch 11/300
1/8, train_loss: 0.5905, step time: 1.5489
2/8, train_loss: 0.5844, step time: 1.5277
3/8, train_loss: 0.6046, step time: 1.5266
4/8, train_loss: 0.5408, step time: 1.5271
5/8, train_loss: 0.5791, step time: 1.5284
6/8, train_loss: 0.5734, step time: 1.5251
7/8, train_loss: 0.5421, step time: 1.5311
8/8, train_loss: 0.6211, step time: 1.5263
epoch 11 average loss: 0.5795

/home/umgpy/Monai_8_test_env/lib/python3.8/site-packages/monai/metrics/hausdorff_distance.py:160: UserWarning: the prediction of class 0 is all 0, this may result in nan/inf distance.
  warnings.warn(f"the prediction of class {c} is all 0, this may result in nan/inf distance.")
/home/umgpy/Monai_8_test_env/lib/python3.8/site-packages/numpy/lib/function_base.py:4486: RuntimeWarning: invalid value encountered in subtract
  diff_b_a = subtract(b, a)
/home/umgpy/Monai_8_test_env/lib/python3.8/site-packages/monai/metrics/hausdorff_distance.py:160: UserWarning: the prediction of class 1 is all 0, this may result in nan/inf distance.
  warnings.warn(f"the prediction of class {c} is all 0, this may result in nan/inf distance.")
/home/umgpy/Monai_8_test_env/lib/python3.8/site-packages/monai/metrics/hausdorff_distance.py:160: UserWarning: the prediction of class 2 is all 0, this may result in nan/inf distance.
  warnings.warn(f"the prediction of class {c} is all 0, this may result in nan/inf distance.")
/home/umgpy/Monai_8_test_env/lib/python3.8/site-packages/monai/metrics/surface_distance.py:155: UserWarning: the prediction of class 0 is all 0, this may result in nan/inf distance.
  warnings.warn(f"the prediction of class {c} is all 0, this may result in nan/inf distance.")
/home/umgpy/Monai_8_test_env/lib/python3.8/site-packages/monai/metrics/surface_distance.py:155: UserWarning: the prediction of class 1 is all 0, this may result in nan/inf distance.
  warnings.warn(f"the prediction of class {c} is all 0, this may result in nan/inf distance.")
/home/umgpy/Monai_8_test_env/lib/python3.8/site-packages/monai/metrics/surface_distance.py:155: UserWarning: the prediction of class 2 is all 0, this may result in nan/inf distance.
  warnings.warn(f"the prediction of class {c} is all 0, this may result in nan/inf distance.")

current epoch: 11 current mean assd: inf BG: inf fem: inf pel: inf
best mean dice: -1.0000 at epoch: -1
time consuming of epoch 11 is: 633.2988

epoch 12/300
1/8, train_loss: 0.6075, step time: 1.5569
2/8, train_loss: 0.6360, step time: 1.5258
3/8, train_loss: nan, step time: 1.5197
4/8, train_loss: 0.6048, step time: 1.5274
5/8, train_loss: 0.6066, step time: 1.5282
6/8, train_loss: 0.5886, step time: 1.5263
7/8, train_loss: nan, step time: 1.5196
8/8, train_loss: 0.5142, step time: 1.5269
epoch 12 average loss: nan
current epoch: 12 current mean assd: inf BG: inf fem: inf pel: inf
best mean dice: -1.0000 at epoch: -1
time consuming of epoch 12 is: 634.2761
```

**Reply by al3ms** [reply]
Hi 
Did you solve the problem? Regarding using Dice loss along with dice metric, and same timeHD95 and average surface distance?

**Comment by dscarmo** [comment]
Writing this response if somebody finds this in the future, distance metrics are not valid when one of the two surfaces is not present in the prediction. 

I am pretty sure the BRATS dataset has some cases where the more severe type of tumor is not present, invalidating surface distance.

In fact the MONAI warning is explaining this, "UserWarning: the prediction of class 1 is all 0, this may result in nan/inf distance."

**Reply by emi-dm** [reply]
@KumoLiu do you know how to solve it?

**Reply by emi-dm** [reply]
In my case, I created a model to segment the liver and liver cancer in MSD Task 03. However, the problem is that, in some cases, the volume labels do not contain a certain class, causing the SurfaceDistanceMetric to return infinity.

**Reply by KumoLiu** [reply]
Hi @emi-dm, if the ground truth does not contain the class, you may skip the calculate. If the pred doesn't contain, you may enhance the model performance.

**Reply by emi-dm** [reply]
This is my code for model validation now @KumoLiu : 

Post Transforms

```
post_label = AsDiscrete(to_onehot=3)
post_trans = Compose([AsDiscrete(argmax=True, to_onehot=3)])
post_trans_to_write = Compose([Activations(softmax=True), AsDiscrete(argmax=True)])
```


Validation in the loop

```
            model.eval()
            eval_loss =  0.0
            val_loader = tqdm(loaders["validation"], desc=f"Epoch {epoch+1}/{epochs} - Validation")
            with torch.no_grad():
                for data in val_loader:
                    inputs, labels = data["image"].to(device), data["label"].to(device)
                    logits = sliding_window_inference(inputs, (128, 128, 32), 10, model, overlap=0.75)
                    loss = criterion(logits, labels)
                    eval_loss += loss.item()
                    outputs = [post_trans(i) for i in decollate_batch(logits)]
                    labels_convert = [post_label(i) for i in decollate_batch(labels)]
                    dice(y_pred=outputs, y=labels_convert)
                    sdm(y_pred=outputs, y=labels_convert)
                    
                eval_dice = dice.aggregate().item()
                eval_sdm = sdm.aggregate().item()
                eval_loss /= len(loaders["validation"])
                print(f"Epoch: {epoch} | Eval Loss: {loss.item()} | Eval Dice: {eval_dice} | Eval SDM: {eval_sdm}")
                dice.reset()
                sdm.reset()
                if eval_dice > best_dice and eval_loss < less_eval_loss:
                    best_dice = eval_dice
                    less_eval_loss = eval_loss

                    # save_dir : Experiment_vN / models
                    val_results_dir = output_dir/f"Epoch_{epoch}"
                    if not val_results_dir.exists():
                        val_results_dir.mkdir(parents=False, exist_ok=True)
                    
                    for i, data in enumerate(loaders["validation"], start=1):
                        inputs, labels = data["image"].to(device), data["label"].to(device)
                        logits = sliding_window_inference(inputs, (128, 128, 32), 10, model, overlap=0.75)
                        outputs = [post_trans_to_write(i) for i in decollate_batch(logits)]
                        # TODO: Compute metrics and save the results in csv file
                        model_weights_dir = save_dir/"best_model.pth"
                        torch.save(model.state_dict(), model_weights_dir)
                        eval_utils.torch_to_nifti(inputs[0], labels[0], outputs[0], epoch=epoch, vol_id=i, output_folder=val_results_dir)
            logger.debug(f"Epoch: {epoch}, Eval Dice: {best_dice}")
```

How could I modify it to avoid possible inf or nan results?
